src/device.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+//! Handles routing to devices in an address space.
+use std::sync::{Arc, Mutex};
+use vm_memory::{GuestAddress, GuestUsize};
+
+/// Trait for devices with basic functions.
+#[allow(unused_variables)]
+pub trait Device: Send {
+    /// Get the device name.
+    fn name(&self) -> String;
+    /// Read from the guest physical address `addr` to `data`.
+    fn read(&mut self, addr: GuestAddress, data: &mut [u8], io_type: IoType);
+    /// Write `data` to the guest physical address `addr`.
+    fn write(&mut self, addr: GuestAddress, data: &[u8], io_type: IoType);
+    /// Set the allocated resource to device.
+    ///
+    /// This will be called by DeviceManager::register_device() to set
+    /// the allocated resource from the vm_allocator back to device.
+    fn set_resources(&mut self, res: &[IoResource], irq: Option<IrqResource>);
+}
+
+/// IO Resource type.
+#[derive(Debug, Copy, Clone)]
+pub enum IoType {
+    /// Port I/O resource.
+    Pio,
+    /// Memory I/O resource.
+    Mmio,
+    /// Non-exit physically backed mmap IO
+    PhysicalMmio,
+}
+
+/// Device resource information.
+#[derive(Debug, Copy, Clone)]
+pub struct IoResource {
+    /// Resource address.
+    pub addr: Option<GuestAddress>,
```

Comment:
Why do we need Option<> here?

Reply:
Because we needs to separate a specified address request and a non-specific address request. The vm-allocator::address::allocate shows the details.

Reply:
Thanks for explanation. Then a helper function is welcomed which assumes that  the `addr` field contains a valid address and returns it.  So callers don't need to deal with the Option<> wrapper.

Reply:
Sounds good! Will add that in update.